### DISCUSSION TITLE
Extending 'Normal Extensions' with 'PolyArg Extensions' & New 'club' Keyword

### DISCUSSION THREAD

**Discussion by sangeethnandakumar** [dotnet/csharplang#7616] [Language Ideas]
In C# we've normal extension methods that can accept one parameter along the "this" keyword

## Existing Extension Support

```csharp
public static class NormalExtensions
{
    public static bool IsAGoodBoy(this string name)
    {
        return name == 'Sangeeth';
    }
}
```

Then use like

```csharp
var result = "Sangeeth".IsAGoodBoy();
if(result)
{
	Console.WriteLine("Hey good boy");
}
```

<hr/>

### The "PolyArg" Concept
The proposed "PolyArg" extension methods would allow extension method signatures to accept multiple arguments using the this keyword. For example:

```csharp
public static class PolyArgExtensions
{
    public static bool IsAGoodBoyUnder24(this string name, this int val)
    {
        return name == "Sangeeth" && val < 24;
    }
}
```

### Prerequisites
- To use this feature, the C# language might introduce a new keyword, `club` (Similar to "typeof," "nameof," and "stackalloc")
- This keyword would combine instances without allocation, generating new type signatures at compile-time.
- Here's how it could be used in "PolyArg" extensions:

```csharp
var result = club("Sangeeth", 24).IsAGoodBoyUnder24();

if (result)
{
    Console.WriteLine("Hey young good boy");
}
```

### Discussion Points
This proposal raises several questions and considerations:

- Do you think the `"club"` keyword is a good idea, and can it unlock potential future language features?
- What could be the potential implications of introducing this feature?
- Are there any conflicts with existing syntax matching that should be addressed?

### Example - Using "Extension with Tuple" vs. Using "Native PolyArg Extensions" Support
To illustrate the benefits of "PolyArg" extensions, consider this comparison between the tuple approach and the proposed "club" keyword approach:

```csharp
var fullName = await FetchNameFromApi();
var age = await FetchAgeFromDatabase();

// Tuple Approach (Allocates)
var result = (fullName, age).IsAGoodBoyUnder24();

// PolyArg Extension Approach using the new 'club' keyword
var result = club(fullName, age).IsAGoodBoyUnder24();
```

> *I'd edited the discussion for better clarity

**Comment by ufcpp** [comment]
```cs
var result = ("Sangeeth", 24).IsAGoodBoyUnder24();
if (result)
{
    Console.WriteLine("Hey young good boy");
}

public static class PolyArgExtensions
{
    public static bool IsAGoodBoyUnder24(this (string name, int val) t)
    {
        return t.name == "Sangeeth" && t.val < 24;
    }
}
```

**Reply by sangeethnandakumar** [reply]
Right. But will it allocate a new Tuple?

**Reply by chrisoverzero** [reply]
`ValueTuple`, in all arities, is a value type.

**Reply by huoyaoyuan** [reply]
Value tuple is a struct and no allocation will happen.

If the extension method is trivial, it can also be inlined and the temporary struct can be treated as two normal variables.

**Comment by chrisoverzero** [comment]
> This proposal raises several questions and considerations:

What’s the benefit?

**Reply by KennethHoff** [reply]
Seems like there's only 1; no allocations. And even that is wrong as the ValueTuple approach is already allocation-free.

**Comment by sangeethnandakumar** [comment]
Ya it makes sense, I'm closing the discussion. Thanks folks
